### DISCUSSION TITLE
keyboard layers and keyboard shortcuts

### DISCUSSION THREAD

**Discussion by suliveevil** [yqrashawn/GokuRakuJoudo#163] [Ideas]
# keyboard layers and keyboard shortcuts

Think of each modifier key combination as a keyboard layer.

```edn
;; modifier keys
;; "cost" and "qwer", easy to remmber right?

;; C  | left_command
;; O  | left_option
;; S  | left_shift
;; T  | left_control

;; Q  | right_command
;; W  | right_control
;; E  | right_option
;; R  | right_shift

```

## native keyboard layers

You have 15 native keyboard layers with 4 modifier keys.
With 30 normal keys, you get 450 possible keyboard shortcuts.

```edn
;; native keyboard layers
;; C
;; O
;; S
;; T
;; CO
;; CS
;; CT
;; OS
;; OT
;; ST
;; COS
;; COT
;; CST
;; OST
;; COST
```

## extended keyboard layers

What if you have 8 modifier keys? 255 layers! This is the power of math/Pascal's Triangle(杨辉三角)!
With 30 normal keys, you get 7650 possible keyboard shortcuts.
Too much to remember, right? This is why you should use hyper layers.

## hyper layers

Think of hyper layers as a simplified version of extended keyboard layers.
With 5 modifier keys, you have 31 layers in total.
With 30 normal keys, you get 930 possible keyboard shortcuts.
It doubles native keyboard layers and shortcuts, meanwhile much simpler than extended keyboard layers.

```edn
;; hyper-mode | hyper layers - simplified extended layers
;; H/QWER + [COST]{0,4} : 16 layers

;; hyper-0
;; QWER

;; hyper-1
;; QWERC
;; QWERO
;; QWERS
;; QWERT

;; hyper-2
;; QWERCO
;; QWERCS
;; QWERCT
;; QWEROS
;; QWEROT
;; QWERST

;; hyper-3
;; QWERCOS
;; QWERCOT
;; QWERCST
;; QWEROST

;; hyper-4
;; QWERCOST

```

**Comment by joshuawagner** [comment]
### I like where this is going but I can't figure out how to cause multiple simultaneous keys to act as a modifier. For example, I'd like to be able to hold `f` and `d` to enter an `hjkl-mode`

Right now I'm using layers with multi-layering. I took that from here:
https://github.com/yqrashawn/yqdotfiles/blob/master/modules/yqrashawn/home-manager/dotfiles/karabiner.edn#L53
For example, I have a tab-mode and then a tab-e-mode that acts on `e` press but only if tab-mode. This means I have to make sure that I press tab before I hit e, which is manageable, I guess.

I'm also using a simlayer for something like backslash + simultaneous jk but I can't figure out how to have multiple keys in the simlayer.

Any help would be much appreciated!

**Reply by evan-liu** [reply]
Hi @joshuawagner, I think you are looking for [duoLayer](https://evan-liu.github.io/karabiner.ts/rules/duo-layer)?
